apollo-federation/src/operation/mod.rs

```diff
@@ -406,6 +505,40 @@ mod selection_map {
                 .extend(other.iter().map(|(k, v)| (k.clone(), v.clone())))
         }
 
+        /// Iterates over the inner map, passing each selection through the mapper. If the mapper
+        /// returns `Some`, the current selection is removed from the map and returned selection is
+        /// inserted in its place. This method is intented from sparce updates to selections which
+        /// update their keys (such as defer normalization). As such, update selections will hold
+        /// the same position as the former selection.
+        ///
+        /// NOTE: The given `mapper` function is given a mutable reference so that the `Selection`
+        /// can recurse if needed. Should any other change be needed, the mapper should clone the
+        /// `Selection` and return the updated copy.
+        pub(crate) fn update_and_reinsert<F>(&mut self, mut mapper: F)
```

Comment:
This method is likely not going to stay. While it is safely used currently, if the returned selection has a key that already exists in the map, that selection is removed. This is fine if the key of the original selection equals the new key, but this causes problems because the new selection could remove a separate selection.

Reply:
Good point. It is not safe 🙈